Update bivalvelarvae.py with habitat settlement

RomainChaput opened the pull request:
I tried to add the ability to settlement particles in a shapefile (representing a suitable environment). When settlement_in_habitat is True, elements are only deactivated if they are old enough to settle and if they are within the boundaries of an element of the shapefile.
To activate in a script add these lines before running OpenDrift:

###############################
# Define habitat
###############################
shp, bins = o.habitat('./habitat/Test_habitat.shp') # Location of the shapefile with the habitat

###############################
# Type of settlement
###############################
o.set_config('drift:settlement_in_habitat', True)
o.set_config('drift:max_age_seconds', max_PLD) # maximum duration of PLD without settlement
o.set_config('drift:min_settlement_age_seconds', min_PLD) # Beginning of the competency period

RomainChaput committed:
Update bivalvelarvae.py with habitat settlement

I tried to add the ability to settlement particles in a shapefile (representing a suitable environment). When settlement_in_habitat is True, elements are only deactivated if they are old enough to settle and if they are within the boundaries of an element of the shapefile.
To activate in a script add these lines before running OpenDrift:

###############################
# Define habitat
###############################
shp, bins = o.habitat('./habitat/Test_habitat.shp') # Location of the shapefile with the habitat

###############################
# Type of settlement
###############################
o.set_config('drift:settlement_in_habitat', True)
o.set_config('drift:max_age_seconds', max_PLD) # maximum duration of PLD without settlement
o.set_config('drift:min_settlement_age_seconds', min_PLD) # Beginning of the competency period

RomainChaput commented on opendrift/models/bivalvelarvae.py:
```diff
@@ -326,7 +357,32 @@ def interact_with_coastline(self,final = False):
+                       in_habitat = pt.within(poly)
+                       if in_habitat == True:
+                           #import pdb; pdb.set_trace()
+                           self.environment.land_binary_mask[old_enough[i]] = 1
```
# This should work better:
 self.environment.land_binary_mask[old_enough[i]] = 6

RomainChaput commented on opendrift/models/bivalvelarvae.py:
```diff
@@ -326,7 +357,32 @@ def interact_with_coastline(self,final = False):
+                           
+           # Deactivate elements that are within a polygon and old enough to settle
+           # ** function expects an array of size consistent with self.elements.lon                
+           self.deactivate_elements((self.environment.land_binary_mask == 1), reason='settled_in_habitat')
```
# This should work better:
self.deactivate_elements((self.environment.land_binary_mask == 6), reason='settled_in_habitat')

RomainChaput committed:
Update bivalvelarvae.py

Changed self.environment.land_binary_mask == 1 to self.environment.land_binary_mask == 6 when particles are within an habitat (land_binary_mask == 1 was conflicting with settlement at the coast).

RomainChaput committed:
Update bivalvelarvae.py

Calvin found that the function seeded_on_land was flagging the particles that stopped moving long after being released. I changed the function with the original version of basemodel.py and it seems to work as intended.

RomainChaput committed:
Update bivalvelarvae.py

Optimization of the habitat module. The use of the MultiPolygon object should speed up the function.

simonweppe commented:
@RomainChaput  - sorry I've missed this one ! I'll have a good look next week.

RomainChaput commented:
I have a few more changes that I need to push. Calvin is going to test the impact of vertical swimming behaviors on the dispersal, so I added a module to do that. I will push it on Monday.
________________________________
From: simonweppe ***@***.***>
Sent: Friday, 25 June 2021 10:59 PM
To: simonweppe/opendrift ***@***.***>
Cc: Romain Chaput ***@***.***>; Mention ***@***.***>
Subject: Re: [simonweppe/opendrift] Update bivalvelarvae.py with habitat settlement (#1)


@RomainChaput<https://apc01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FRomainChaput&data=04%7C01%7Cromain.chaput%40vuw.ac.nz%7C8c799312f8c745f76bb808d937c85a92%7Ccfe63e236951427e8683bb84dcf1d20c%7C0%7C0%7C637602156102889197%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=D5hcSxIwZboNFRAtgJnNXe%2BQt%2FOhQ8ahNaXWYk%2FVZx0%3D&reserved=0> - sorry I've missed this one ! I'll have a good look next week.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://apc01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fsimonweppe%2Fopendrift%2Fpull%2F1%23issuecomment-868417635&data=04%7C01%7Cromain.chaput%40vuw.ac.nz%7C8c799312f8c745f76bb808d937c85a92%7Ccfe63e236951427e8683bb84dcf1d20c%7C0%7C0%7C637602156102899174%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=GaqRYUs3tEHUhHnPvP9XVht%2BSp%2BKTBYD2aNUYS0hmHQ%3D&reserved=0>, or unsubscribe<https://apc01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FASE3RI6FKBLHZ2VHAVKAFETTUROSHANCNFSM43YXBVXA&data=04%7C01%7Cromain.chaput%40vuw.ac.nz%7C8c799312f8c745f76bb808d937c85a92%7Ccfe63e236951427e8683bb84dcf1d20c%7C0%7C0%7C637602156102904171%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=xLVNnThjIIuoUKHtwZjGaqeFmRYleXs0Au1GoElcrbk%3D&reserved=0>.

RomainChaput committed:
Update of bivalvelarvae.py for habitat settlement, vertical swimming, maximum depth, and Haliotis iris

I added a vertical swimming behavior correlated to the direction of the currents: when larvae are advected away from the coast they start a correlated random walk, swimming up and down until they find a current that bring them closer to shore. Because this code involves calling the habitat module at every timestep, I used a Ball Tree algorithm to classify the habitat centroids and improve the efficiency of the searching algorithm.
I also added a maximum dispersal depth that the user can specify.
Finally, I added an option to simulate the vertical migration of Haliotis iris during the first 12 hours of dispersal (sink, then rise to the surface).
